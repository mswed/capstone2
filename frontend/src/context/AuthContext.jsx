import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
