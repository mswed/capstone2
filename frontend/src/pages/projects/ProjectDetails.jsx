import { useState, useEffect, useContext } from 'react';
import { useParams } from 'react-router-dom';
import { Container, Card, Row, Col } from 'react-bootstrap';
import { AuthContext } from '../../context/AuthContext.jsx';
import GrumpyApi from '../../services/api.js';
import Loading from '../../components/ui/Loading.jsx';
import CameraGrid from '../../features/cameras/components/CamerasGrid.jsx';
import FormatList from '../../features/formats/components/FormatList.jsx';
import ActionBar from '../../components/ui/ActionBar.jsx';
import FormatSearchModal from '../../features/formats/components/FormatSearchModal.jsx';

const ProjectDetails = () => {
  // Set up state
  const { projectId } = useParams();
  const [projectData, setProjectData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [showFormatsModal, setShowFormatsModal] = useState(false);
  const { token, currentUser } = useContext(AuthContext);

  const actionButtons = [
    {
      text: 'Add Format',
      variant: 'outline-primary',
      onClick: () => setShowFormatsModal(true),
    },
  ];

  const handleAddFormat = async (formatId) => {
    try {
      await GrumpyApi.addFormatToProject(projectId, formatId);
      // Refresh the project
      const response = await GrumpyApi.getProjectDetails(projectId);
      setProjectData(response);
    } catch (error) {}
  };

  const handleVote = async (formatId, vote) => {
    try {
      await GrumpyApi.voteOnProjectFormat(projectId, formatId, vote, currentUser);
      // Refresh the project
      const response = await GrumpyApi.getProjectDetails(projectId);
      setProjectData(response);
    } catch (error) {
      console.error('Error failed to vote on format', error);
    }
  };

  // Fetch project data
  useEffect(() => {
    const getProjectDetails = async () => {
      try {
        const response = await GrumpyApi.getProjectDetails(projectId);
        setProjectData(response);
      } catch (error) {
        console.error('Error fetching camera details', error);
      } finally {
        setIsLoading(false);
      }
    };
    getProjectDetails();
  }, [projectId]);

  if (isLoading) {
    return <Loading />;
  }

  return (
    <Container>
      <Card className="shadow-lg rounded-0 rounded-bottom">
        <Row className="g-0">
          <Col md={4}>
            <Card.Img
              src={projectData.poster_path ? projectData.poster_path : '/media/camera_images/missing_image.png'}
              alt={`Poster for for ${projectData.name}`}
              className="p-3"
              style={{
                objectFit: 'contain',
                maxHeight: '350px',
                maxWidth: '100%',
              }}
            />
          </Col>
          <Col md={5}>
            <Card.Body className="text-start">
              <Card.Title>
                <h2>
                  {projectData.name} ({projectData.release_date.split('-')[0]})
                </h2>
              </Card.Title>

              <div className="mt-3">
                <h4>Description</h4>
                <p>{projectData.description} </p>
              </div>
              <CameraGrid cameras={projectData.cameras} />
            </Card.Body>
          </Col>
          <Col md={3} className="text-start p-3">
            <div>
              <dl className="row">
                <dt className="col-sm-4">Type:</dt>
                <dd className="col-sm-8">{projectData.project_type}</dd>
              </dl>
            </div>
          </Col>
        </Row>
      </Card>
      {token && <ActionBar buttons={actionButtons} className="mt-3" />}
      <FormatSearchModal show={showFormatsModal} onHide={() => setShowFormatsModal(false)} onFormatSelect={handleAddFormat} projectId={projectId} />
      {projectData.formats?.length > 0 && (
        <Row className="mt-3">
          <div className="text-start">
            <h3>Formats</h3>
          </div>
          <FormatList formats={projectData.formats} showModel={true} onVote={handleVote} />
        </Row>
      )}
    </Container>
  );
};

export default ProjectDetails;
