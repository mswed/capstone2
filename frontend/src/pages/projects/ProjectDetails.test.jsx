import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import ProjectDetails from './ProjectDetails.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';
import { MessagesProvider } from '../../context/MessageContext.jsx';
import GrumpyApi from '../../services/api.js';

// When adding a format to a project we need some format functions like
// scrollTo
global.scrollTo = vi.fn();

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getProjectDetails: vi.fn(() =>
      Promise.resolve({
        id: 1,
        name: 'One Piece',
        projectType: 'episodic',
        description: 'With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure.',
        posterPath: 'https://www.pathtoposter1.com',
        releaseDate: '2023-08-31',
        adult: false,
        tmdb_id: 123456,
        tmdbOriginalName: 'same as the other name',
        genres: ['adventure'],
        rating: 'PG-13',
        cameras: [
          {
            id: 1,
            make: 1,
            makeName: 'ARRI',
            model: 'Alexa 35',
            sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
            sensorSize: 'Super 35',
            maxFilmbackWidth: 27.99,
            maxFilmbackHeight: 19.22,
            maxImageWidth: 4608,
            maxImageHeight: 3164,
            minFrameRate: 0.75,
            maxFrameRate: 120,
            notes: 'sample notes',
          },
        ],
        formats: [
          {
            id: 1,
            camera: 1,
            cameraModel: 'Alexa 35',
            makeName: 'ARRI',
            imageFormat: '4.6K',
            imageAspect: '3:2',
            formatName: 'Open Gate',
            sensorWidth: 28.0,
            sensorHeight: 19.2,
            imageWidth: 4608,
            imageHeight: 3164,
            pixelAspect: 1.0,
            isAnamorphic: false,
            isDesqueezed: false,
            anamorphicSqueeze: null,
            filmbackWidth3de: 28.0,
            filmbackHeight3de: 19.2,
            isDownsampled: false,
            isUpscaed: false,
            codec: 'ARRIRAW',
            rawReordingAvailable: true,
            source: 1,
            notes: 'sample notes',
            makeNotes:
              '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
            trackingWorkflow: 'Do this then that',
            upVotes: 3,
            downVotes: 0,
            totalVotes: 3,
            score: 3,
            userVote: 'up',
          },
          {
            id: 2,
            camera: 1,
            cameraModel: 'Alexa 35',
            makeName: 'ARRI',
            imageFormat: '4K',
            imageAspect: '16:9',
            sensorWidth: 28.0,
            sensorHeight: 15.7,
            imageWidth: 4096,
            imageHeight: 2304,
            pixelAspect: 1.0,
            isAnamorphic: false,
            isDesqueezed: false,
            anamorphicSqueeze: null,
            filmbackWidth3de: 28.0,
            filmbackHeight3de: 19.2,
            isDownsampled: true,
            isUpscaed: false,
            codec: 'ProRes',
            rawReordingAvailable: true,
            source: 1,
            notes: 'Downsampled from 4.6K 16:9',
            makeNotes:
              '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
            trackingWorkflow: 'Do that then this',
            upVotes: 0,
            downVotes: 2,
            totalVotes: 2,
            score: -2,
            userVote: 'down',
          },
        ],
      })
    ),
    addFormat: vi.fn(() =>
      Promise.resolve({
        success: 'Partialy updated format',
        format: {
          id: 1,
          camera: 1,
          cameraModel: 'Alexa 35 Updated',
          makeName: 'ARRI',
          imageFormat: '4.6K',
          imageAspect: '3:2',
          formatName: 'Open Gate',
          sensorWidth: 28.0,
          sensorHeight: 19.2,
          imageWidth: 4608,
          imageHeight: 3164,
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: false,
          isUpscaed: false,
          codec: 'ARRIRAW',
          rawReordingAvailable: true,
          source: 1,
          notes: 'sample notes',
          makeNotes:
            '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
          trackingWorkflow: 'Do this then that',
        },
      })
    ),
    deleteProject: vi.fn(() =>
      Promise.resolve({
        success: 'Project deleted',
      })
    ),
    findFormats: vi.fn(() =>
      Promise.resolve([
        {
          id: 3,
          camera: 3,
          cameraModel: 'KOMODO',
          makeName: 'RED',
          imageFormat: '6K',
          imageAspect: '2.4:1',
          sensorWidth: 27.03,
          sensorHeight: 11.4,
          imageWidth: 6144,
          imageHeight: 2592,
          codec: 'R3D',
          source: 5,
          notes: 'There is a discrapency in the documentation the height is either 2592 or 2574',
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: true,
          isUpscaed: false,
          rawReordingAvailable: true,
          makeNotes: '10 out of 10, no notes',
          trackingWorkflow: 'Do that then this',
        },
      ])
    ),

    getFormats: vi.fn(() =>
      Promise.resolve([
        {
          id: 1,
          camera: 1,
          cameraModel: 'Alexa 35',
          makeName: 'ARRI',
          imageFormat: '4.6K',
          imageAspect: '3:2',
          formatName: 'Open Gate',
          sensorWidth: 28.0,
          sensorHeight: 19.2,
          imageWidth: 4608,
          imageHeight: 3164,
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: false,
          isUpscaed: false,
          codec: 'ARRIRAW',
          rawReordingAvailable: true,
          source: 1,
          notes: 'sample notes',
          makeNotes:
            '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
          trackingWorkflow: 'Do this then that',
        },
        {
          id: 2,
          camera: 1,
          cameraModel: 'Alexa 35',
          makeName: 'ARRI',
          imageFormat: '4K',
          imageAspect: '16:9',
          sensorWidth: 28.0,
          sensorHeight: 15.7,
          imageWidth: 4096,
          imageHeight: 2304,
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: true,
          isUpscaed: false,
          codec: 'ProRes',
          rawReordingAvailable: true,
          source: 1,
          notes: 'Downsampled from 4.6K 16:9',
          makeNotes:
            '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
          trackingWorkflow: 'Do that then this',
        },
      ])
    ),
  },
}));

describe('ProjectDetails component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter initialEntries={['/projects/1']}>
        <MessagesProvider>
          <AuthContext.Provider value={authValue}>
            <Routes>
              <Route path="/projects/:projectId" element={<ProjectDetails />} />
            </Routes>
          </AuthContext.Provider>
        </MessagesProvider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findByText('One Piece (2023)');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findByText('One Piece (2023)');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();

    await waitFor(() => {
      expect(screen.getByText('One Piece (2023)')).toBeInTheDocument();
      expect(screen.getByText('With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure.')).toBeInTheDocument();
      expect(screen.getByText('episodic')).toBeInTheDocument();
      // expect(screen.getByText('4.6K 3:2 Open Gate')).toBeInTheDocument();
    });
  });

  it('shows action bar for admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'admminUser',
      isAdmin: true,
    });
    await screen.findByText('One Piece (2023)');
    expect(screen.getByText('Delete Project')).toBeInTheDocument();
  });

  it('shows action bar for regular users', async () => {
    renderWithAuth({ token: 'fake-token', currentUser: 'regularUser' });
    await screen.findByText('One Piece (2023)');
    expect(screen.queryByText('Delete Project')).not.toBeInTheDocument();
    expect(screen.getByText('Add Format')).toBeInTheDocument();
  });
});
