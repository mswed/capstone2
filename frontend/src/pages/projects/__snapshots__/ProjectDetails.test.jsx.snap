// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="shadow-lg rounded-0 rounded-bottom card"
    >
      <div
        class="g-0 row"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="Poster for for One Piece"
            class="card-img p-3"
            src="https://www.pathtoposter1.com"
            style="object-fit: contain; max-height: 350px; max-width: 100%;"
          />
        </div>
        <div
          class="col-md-5"
        >
          <div
            class="text-start card-body"
          >
            <div
              class="card-title h5"
            >
              <h2>
                One Piece (2023)
              </h2>
            </div>
            <div
              class="mt-3"
            >
              <h4>
                Description
              </h4>
              <p>
                With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure. 
              </p>
            </div>
            <div
              class="row"
            >
              <div
                class="px-3 col-md-4 col-sm-6 col-12"
              >
                <a
                  class="text-decoration-none"
                  data-discover="true"
                  href="/cameras/1"
                >
                  <div
                    class="h-50 shadow-lg card"
                  >
                    <img
                      alt="Image for Alexa 35"
                      class="card-img-top w-50 mx-auto mt-3"
                      style="object-fit: contain;"
                    />
                    <div
                      class="card-body"
                    >
                      <p
                        class="fs-6 text-center mb-1 card-text"
                      >
                         Alexa 35
                      </p>
                    </div>
                  </div>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="text-start p-3 col-md-3"
        >
          <div>
            <dl
              class="row"
            >
              <dt
                class="col-sm-4"
              >
                Type:
              </dt>
              <dd
                class="col-sm-8"
              >
                episodic
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
    <div
      class="border-0 shadow-sm mt-3 card"
    >
      <div
        class="py-2 card-body"
      >
        <div
          class="align-items-center row"
        >
          <div
            class="text-end col"
          >
            <button
              class="m-2  btn btn-outline-primary btn-sm"
              type="button"
            >
              Add Format
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-3 row"
    >
      <div
        class="text-start"
      >
        <h3>
          Formats
        </h3>
      </div>
      <div
        class="table-responsive"
      >
        <table
          class="small table table-sm table-striped table-hover"
        >
          <thead>
            <tr>
              <th>
                Make
              </th>
              <th>
                Model
              </th>
              <th>
                Format
              </th>
              <th>
                Resolution
              </th>
              <th>
                Filmback
              </th>
              <th>
                PAR
              </th>
              <th>
                Anamorphic?
              </th>
              <th>
                Desqueezed?
              </th>
              <th />
              <th>
                Action
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                ARRI
              </td>
              <td>
                Alexa 35
              </td>
              <td>
                4.6K 3:2 Open Gate
              </td>
              <td>
                4608 x 3164
              </td>
              <td>
                28mm x 19.2mm
              </td>
              <td>
                1
              </td>
              <td
                class="text-center"
              >
                <span
                  class="text-danger"
                >
                  ✗
                </span>
              </td>
              <td
                class="text-center"
              >
                <span
                  class="text-danger"
                >
                  ✗
                </span>
              </td>
              <td>
                <div
                  class="d-flex flex-column align-items-center"
                  style="min-width: 60px;"
                >
                  <button
                    class="btn p-0 border-0"
                    style="color: rgb(40, 167, 69); font-size: 18px; line-height: 1;"
                  >
                    ▲
                  </button>
                  <small
                    class="text-muted my-1"
                  >
                    3
                  </small>
                  <button
                    class="btn p-0 border-0"
                    style="color: rgb(220, 53, 69); font-size: 18px; line-height: 1;"
                  >
                    ▽
                  </button>
                </div>
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/formats/1"
                >
                  View
                </a>
              </td>
            </tr>
            <tr>
              <td>
                ARRI
              </td>
              <td>
                Alexa 35
              </td>
              <td>
                4K 16:9 
              </td>
              <td>
                4096 x 2304
              </td>
              <td>
                28mm x 15.7mm
              </td>
              <td>
                1
              </td>
              <td
                class="text-center"
              >
                <span
                  class="text-danger"
                >
                  ✗
                </span>
              </td>
              <td
                class="text-center"
              >
                <span
                  class="text-danger"
                >
                  ✗
                </span>
              </td>
              <td>
                <div
                  class="d-flex flex-column align-items-center"
                  style="min-width: 60px;"
                >
                  <button
                    class="btn p-0 border-0"
                    style="color: rgb(40, 167, 69); font-size: 18px; line-height: 1;"
                  >
                    △
                  </button>
                  <small
                    class="text-muted my-1"
                  >
                    -2
                  </small>
                  <button
                    class="btn p-0 border-0"
                    style="color: rgb(220, 53, 69); font-size: 18px; line-height: 1;"
                  >
                    ▼
                  </button>
                </div>
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/formats/2"
                >
                  View
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
